### PR TITLE
update dm-ansible, support ha

### PR DESCRIPTION
**Pull request by liubo0127**
### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@csuzhangxc @WangXiangUSTC PTAL, ref [DM-ansible 兼容 HA](https://docs.google.com/document/d/1q-puatdA65ZtSp0iOVneEbS2MPnLGpRv_ZCJCR5uXpM) (internal only).

File: dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2 [hunk 1]
```diff
@@ -5,12 +5,26 @@ ulimit -n 1000000
 DEPLOY_DIR={{ deploy_dir }}
 cd "${DEPLOY_DIR}" || exit 1
 
+{% set my_ip = hostvars[inventory_hostname].ansible_host | default(hostvars[inventory_hostname].inventory_hostname) -%}
+{% set my_port = hostvars[inventory_hostname].dm_worker_port -%}
+{% set my_separator = "_" %}
+{% set my_hostname = hostvars[inventory_hostname]['ansible_hostname'] | default(hostvars[inventory_hostname].inventory_hostname) -%}
+{% set my_peer_id = my_separator ~ my_hostname ~ my_separator ~ my_port %}
+
+{% set all_dm_worker = [] -%}
```
Comment (csuzhangxc) [inline dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2]: Is naming it as `all_dm_master` more accurate?

File: dm/dm-ansible/rolling_update.yml [hunk 2]
```diff
@@ -10,62 +10,79 @@
...
-    - name: wait for dm-worker up
+    - name: wait for dm-master up
       wait_for: |
-        host={{ ansible_host }} port={{ dm_worker_port }} state=started
-
+        host={{ ansible_host }} port={{ dm_master_port }} state=started
+    
+    - name: wait until dm_master is available
+      uri:
+        url: "http://{{ ansible_host }}:{{ dm_master_port }}/apis/v1alpha1/status"
```
Comment (csuzhangxc) [inline dm/dm-ansible/rolling_update.yml]:
```suggestion
        url: "http://{{ ansible_host }}:{{ dm_master_port }}/apis/v1alpha1/status/"
```

File: dm/dm-ansible/scripts/dm.json [hunk 3]
```diff
@@ -3360,6 +3619,7 @@
     ]
   },
   "timezone": "",
-  "title": "DM-task",
-  "version": 28
+  "title": "Test-Cluster-DM-worker",
```
Comment (csuzhangxc) [inline dm/dm-ansible/scripts/dm.json]:
```suggestion
  "title": "Test-Cluster-DM-task",
```

File: dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2 (hunk 1 above)
Comment (liubo0127) [reply dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2]: `all_dm_master` is in `dm/dm-ansible/roles/dm-worker/templates/run_dm-master.sh.j2`

Comment (csuzhangxc) [reply dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2]: these are `all_dm_master` which `dm_worker` is joining.

Comment (liubo0127) [reply dm/dm-ansible/roles/dm-worker/templates/run_dm-worker.sh.j2]: Make sense, agree

**Comment by csuzhangxc** [comment]
when I run `ansible-playbook start.yml`, I got

```
TASK [import grafana dashboards - run import script] ************************************************
fatal: [grafana -> localhost]: FAILED! => {"changed": true, "cmd": "python import_grafana_dashboards.py", "delta": "0:00:00.336520", "end": "2020-03-03 17:31:38.703337", "msg": "non-zero return code", "rc": 1, "start": "2020-03-03 17:31:38.366817", "stderr": "Traceback (most recent call last):\n  File \"import_grafana_dashboards.py\", line 79, in <module>\n    dashboard = fill_dashboard_with_dest_config(dashboard, dest, type_)\n  File \"import_grafana_dashboards.py\", line 38, in fill_dashboard_with_dest_config\n    for row in dashboard['rows']:\nKeyError: u'rows'", "stderr_lines": ["Traceback (most recent call last):", "  File \"import_grafana_dashboards.py\", line 79, in <module>", "    dashboard = fill_dashboard_with_dest_config(dashboard, dest, type_)", "  File \"import_grafana_dashboards.py\", line 38, in fill_dashboard_with_dest_config", "    for row in dashboard['rows']:", "KeyError: u'rows'"], "stdout": "[load] from <dm_instances.json>:dm_worker_instances", "stdout_lines": ["[load] from <dm_instances.json>:dm_worker_instances"]}
```

File: dm/dm-ansible/roles/dm-master/defaults/main.yml [hunk 4]
```diff
@@ -5,5 +5,7 @@
 dm_master_log_level: "info"
 dm_master_log_dir: "{{ deploy_dir }}/log"
 dm_master_log_file: "dm-master.log"
+dm_master_data_dir: "{{ deploy_dir }}/data.dm_master"
 dm_master_stdout_log_file: "dm-master-stdout.log"
 dm_master_stderr_log_file: "dm-master-stderr.log"
+dm_master_name_prefix: "dm_master"
```
Comment (csuzhangxc) [inline dm/dm-ansible/roles/dm-master/defaults/main.yml]: we often name the dm-master in `inventory.ini` as `dm_master`, after added this prefix, the name will become `dm_master_dm_master_8261`, is this a bit verbose?

**Comment by liubo0127** [comment]
> when I run `ansible-playbook start.yml`, I got
> 
> ```
> TASK [import grafana dashboards - run import script] ************************************************
> fatal: [grafana -> localhost]: FAILED! => {"changed": true, "cmd": "python import_grafana_dashboards.py", "delta": "0:00:00.336520", "end": "2020-03-03 17:31:38.703337", "msg": "non-zero return code", "rc": 1, "start": "2020-03-03 17:31:38.366817", "stderr": "Traceback (most recent call last):\n  File \"import_grafana_dashboards.py\", line 79, in <module>\n    dashboard = fill_dashboard_with_dest_config(dashboard, dest, type_)\n  File \"import_grafana_dashboards.py\", line 38, in fill_dashboard_with_dest_config\n    for row in dashboard['rows']:\nKeyError: u'rows'", "stderr_lines": ["Traceback (most recent call last):", "  File \"import_grafana_dashboards.py\", line 79, in <module>", "    dashboard = fill_dashboard_with_dest_config(dashboard, dest, type_)", "  File \"import_grafana_dashboards.py\", line 38, in fill_dashboard_with_dest_config", "    for row in dashboard['rows']:", "KeyError: u'rows'"], "stdout": "[load] from <dm_instances.json>:dm_worker_instances", "stdout_lines": ["[load] from <dm_instances.json>:dm_worker_instances"]}
> ```

fixed

**Comment by csuzhangxc** [comment]
@liubo0127 one more question, it seems `downloads/grafana-6.1.6.tar.gz` not extract to `resources/bin/`.

**Comment by csuzhangxc** [comment]
@WangXiangUSTC PTAL

**Pull request merged**